python_modules/dagster/dagster_tests/core_tests/asset_defs_tests/test_assets_job.py

```diff
@@ -637,6 +638,80 @@ def bar(foo):
         job.execute_in_process()
 
 
+def test_internal_asset_deps():
+    with pytest.raises(Exception, match="output_name non_exist_output_name"):
+
+        @op
+        def my_op(x, y):
+            return x
+
+        @graph(ins={"x": GraphIn()})
+        def my_graph(x, y):
+            my_op(x, y)
+
+        assets_def = AssetsDefinition.from_graph(
+            graph_def=my_graph, internal_asset_deps={"non_exist_output_name": {AssetKey("b")}}
+        )
+
+
+def test_asset_def_from_graph_inputs():
+    @op
+    def my_op(x, y):
+        return x
+
+    @graph(ins={"x": GraphIn()})
+    def my_graph(x, y):
+        my_op(x, y)
+
+    assets_def = AssetsDefinition.from_graph(
+        graph_def=my_graph,
+        asset_keys_by_input_name={"x": AssetKey("x_asset")},
+    )
+
+    assert assets_def.asset_keys_by_input_name["x"] == AssetKey("x_asset")
+    assert assets_def.asset_keys_by_input_name["y"] == AssetKey("y")
+
+
+def test_asset_def_from_graph_outputs():
+    @op
+    def x_op(x):
+        return x
+
+    @op
+    def y_op(y):
+        return y
+
+    @graph(out={"x": GraphOut(), "y": GraphOut()})
+    def my_graph(x, y):
+        return {"x": x_op(x), "y": y_op(y)}
+
+    assets_def = AssetsDefinition.from_graph(
+        graph_def=my_graph,
+        asset_keys_by_output_name={"y": AssetKey("y_asset")},
```

Comment:
I think it might be best to keep asset_keys_by_output_name as an "all or nothing" type of argument, where you either supply every single output -> asset key mapping (and none are inferred), or you leave this argument blank (and all of them are inferred).

This isn't a super hardline stance, but I think this generally reduces the likelihood of silent user errors, where they add in a new output to the graph but forget to map it to an asset key (and instead we silently generate one for them). I understand the convenience argument (sometimes you just want to something special with one of the outputs), but I think in most cases you're either going to be happy with whatever defaults we have, or not (and in that case want to do everything custom).

Reply:
same goes for asset_keys_by_input_name

Reply:
Gotcha. I agree with this, and its easier later toggle this functionality if we have these checks. I've changed the implementation to assert that the set of output/input names specified must equal the output/input names of the graph.